rust/src/model.rs

```diff
@@ -0,0 +1,569 @@
+        &self.weights
+    }
+
+    /// Cast the weights as ratios.
+    fn as_ratios(&self) -> Vec<Ratio<BigInt>> {
+        self.weights
+            .iter()
+            .map(|weight| Ratio::from_integer(BigInt::from(*weight)))
+            .collect()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::iter;
+
+    use rand::distributions::{Distribution, Uniform};
+
+    use super::*;
+    use crate::mask::{
+        config::{
+            BoundType::{Bmax, B0, B2, B4, B6},
+            DataType::{F32, F64, I32, I64},
+            GroupType::{Integer, Power2, Prime},
+            MaskConfigs,
+            ModelType::{M12, M3, M6, M9},
+        },
+        seed::MaskSeed,
+        MaskIntegers,
+    };
+
+    #[test]
+    fn test_model_f32() {
+        // try from
+        let weights = vec![-1_f32, 0_f32, 1_f32];
+        let neg_inf_weights = vec![f32::NEG_INFINITY, 0_f32, 1_f32];
+        let nan_weights = vec![-1_f32, f32::NAN, 1_f32];
+        let inf_weights = vec![-1_f32, 0_f32, f32::INFINITY];
+        let mut model = Model::<f32>::try_from(weights.clone()).unwrap();
+        assert_eq!(model.weights(), &weights);
+        assert_eq!(
+            Model::<f32>::try_from(inf_weights.clone()).unwrap_err(),
+            PetError::InvalidModel,
+        );
+        assert_eq!(
+            Model::<f32>::try_from(neg_inf_weights.clone()).unwrap_err(),
+            PetError::InvalidModel,
+        );
+        assert_eq!(
+            Model::<f32>::try_from(nan_weights.clone()).unwrap_err(),
+            PetError::InvalidModel,
+        );
+
+        // as ratio
+        assert_eq!(
+            model.as_ratios(),
+            vec![
+                Ratio::<BigInt>::from_float(-1_f32).unwrap(),
+                Ratio::<BigInt>::zero(),
+                Ratio::<BigInt>::from_float(1_f32).unwrap(),
+            ],
+        );
+        model.weights = neg_inf_weights;
+        assert_eq!(
+            model.as_ratios(),
+            vec![
+                Ratio::<BigInt>::from_float(f32::MIN).unwrap(),
+                Ratio::<BigInt>::zero(),
+                Ratio::<BigInt>::from_float(1_f32).unwrap(),
+            ],
+        );
+        model.weights = nan_weights;
+        assert_eq!(
+            model.as_ratios(),
+            vec![
+                Ratio::<BigInt>::from_float(-1_f32).unwrap(),
+                Ratio::<BigInt>::zero(),
+                Ratio::<BigInt>::from_float(1_f32).unwrap(),
+            ],
+        );
+        model.weights = inf_weights;
+        assert_eq!(
+            model.as_ratios(),
+            vec![
+                Ratio::<BigInt>::from_float(-1_f32).unwrap(),
+                Ratio::<BigInt>::zero(),
+                Ratio::<BigInt>::from_float(f32::MAX).unwrap(),
+            ],
+        );
+    }
+
+    #[test]
+    fn test_model_f64() {
+        // try from
+        let weights = vec![-1_f64, 0_f64, 1_f64];
+        let neg_inf_weights = vec![f64::NEG_INFINITY, 0_f64, 1_f64];
```

Comment:
I think it would make sense to split these cases in different test functions, it wouldn't add much code, and would make the tests simpler to read.

Reply:
i'll split it.